src/mlpack/core/optimizers/sdp/lrsdp_function_impl.hpp

```diff
@@ -96,6 +98,9 @@ UpdateObjective(double& objective,
   for (size_t i = 0; i < ais.size(); ++i)
   {
     // Take the trace subtracted by the b_i.
+    // Here taking AR first is not recommended as we are already
+    // using pre-computed RR^T. Taking AR first will result in increase
+    // in number of computations.
```

Comment:
I think it would be possible to refactor this in such a way that the `coordinates` matrix is passed directly to `UpdateObjective()`.  Do you think that this would provide additional speedup?

Also, a minor note, but I think that it would be more clear to write `A*R` instead of `AR` in the comments (and the same for the other matrix expression).  It's your call whether or not to make that change; I'm not picky.

Reply:
@rcurtin If I am understanding correctly then here refactoring means to directly pass coordinates matrix inplace of RR^T and then compute the constraint using A*R^T. If this is correct then as per my observation it will increase the number of computations drastically as in current state we are only doing computation in cumulative sum `accu()` and element-wise multiplication `ais[i] % rrt` (_here RR^T is being computed only once in the calling function_) where as using coordinates matrix we will be computing `trans((R^T * A_i) * R)` for each iteration 1 to `ais.size()` which will result in a large increase of computations w.r.t to previous one.
```
UpdateObjective(double& objective,
                const arma::mat& coordinates,
                const std::vector<MatrixType>& ais,
                const arma::vec& bis,
                const arma::vec& lambda,
                const size_t lambdaOffset,
                const double sigma)
{
  for (size_t i = 0; i < ais.size(); ++i)
  {
    // Take the trace subtracted by the b_i.
    const double constraint = trace((trans(coordinates) * ais[i]) * coordinates) - bis[i];
    objective -= (lambda[lambdaOffset + i] * constraint);
    objective += (sigma / 2.) * constraint * constraint;
  }
}
```
I have tried with the above code block and it increased the running time by about 6 times of the current time.
Please correct me if I am going wrong somewhere.

Reply:
So, it's worth pointing out that in the tests you're running, the only functions actually being called are `EvaluateImpl()` and `GradientImpl()` (because the specializations for the `AugLagrangian` optimizer only call those two functions, not the `Evaluate()` and `Gradient()` functions).  So I would have expected that changing to `(A*R)*R^T` for each `A` would be faster, if `(C*R)*R^T` gives such a significant speedup.

The whole speedup depends on this expression being optimized such that the full matrix is not calculated:

```
trace((trans(coordinates) * ais[i]) * coordinates)
```

I'm pretty certain that is true with dense matrices, but I am not sure it is true with sparse matrices.  However, we do have one test case with all dense A matrices: `LRSDPTest/GaussianMatrixSensingSDP`.  Would you be willing to test the three versions of the code (original, optimized C calculation, optimized ais[i] calculations) on that problem?  I suspect we might find that the problem is that when `ais[i]` is sparse, Armadillo is computing the full matrix `trans(coordinates) * ais[i]) * coordinates` as a temporary, then taking the trace of it---which is probably the cause of the slowdown.

Now, I may also be wrong, but my intuition still suggests that if this is faster for the C matrix calculation, we should be able to get the same speedup for the A matrix calculations.

Reply:
Sure I will shortly provide the results for all three conditions.